### PR TITLE
feat: better signature verification + upd mods

### PR DESCRIPTION
**Pull request by omusil24**

### CODE REVIEW

File: crypt/signature.go [hunk 1]
```diff
@@ -35,17 +35,57 @@ func SignWithTimestamp(privateKey *ecdsa.PrivateKey, message string) (string, st
...
-	publicKeyECDSA, err := crypto.SigToPub(crypto.Keccak256Hash([]byte(message)).Bytes(), common.FromHex(signature))
+	hash := crypto.Keccak256([]byte(message))
+	publicKeyECDSA, err := crypto.SigToPub(hash, common.FromHex(signature))
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	publicKeyBytes := crypto.FromECDSAPub(publicKeyECDSA)
+	signatureBytes := common.FromHex(signature)[:64]
+	if !crypto.VerifySignature(publicKeyBytes, hash, signatureBytes) {
```
Comment (shuning-auki) [inline crypt/signature.go]: this is unnecessary.

File: crypt/signature.go [hunk 2]
```diff
@@ -35,17 +35,57 @@ func SignWithTimestamp(privateKey *ecdsa.PrivateKey, message string) (string, st
...
+
+	hash := crypto.Keccak256([]byte(message))
+	signatureBytes := common.FromHex(signature)[:64]
+	publicKeyBytes := crypto.FromECDSAPub(publicKey)
+
+	if !crypto.VerifySignature(publicKeyBytes, hash, signatureBytes) {
+		return common.Address{}, errors.New("signature is incorrect")
+	}
+
+	publicKeyECDSA, err := crypto.SigToPub(hash, common.FromHex(signature))
```
Comment (shuning-auki) [inline crypt/signature.go]: this is unnecessary cause the public key is already in the parameters

**Review by shuning-auki — COMMENTED** [review]
I dont think this PR helps anyone. There is nothing wrong with the original `ValidateSignedMessage`. HDS would stick to `ValidateSignedMessage` cause Hagall doesn't send wallet address during registration. DDS probably won't use `VerifySignedMessage` neither, because as far as I know it is not possible to convert a wallet address into an ecdsa.PublicKey...

**Comment by omusil24** [comment]
> I dont think this PR helps anyone. There is nothing wrong with the original `ValidateSignedMessage`. HDS would stick to `ValidateSignedMessage` cause Hagall doesn't send wallet address during registration. DDS probably won't use `VerifySignedMessage` neither, because as far as I know it is not possible to convert a wallet address into an ecdsa.PublicKey...

Both should be changed to use Verify. Both Hagall and DS should send public key during registration.

Edit: ValidateSignedMessage is broken because it succeeds for different messages than the one used for the signature.

File: crypt/signature.go (hunk 1 above)
Comment (omusil24) [reply crypt/signature.go]: This should be there. But anyway, this function will be removed later on.

**Review by tatiesmars — DISMISSED** [review]
The nice thing with receiving the public key from the node (even if in theory, it is redundant data), it will help us catch bug if a client are tampering the message...
I dont think we need to add VerifySignature in ValidateSignedMessage as it will always return true

**Review by tatiesmars — APPROVED** [review]

**Pull request merged**